File: Badges.swift
//
//  Badges.swift
//  Beerify
//
//  Achievements you can earn across nights. Evaluated locally at the end of
//  each night. Badge state is a set of IDs in AppData.unlockedBadges.
//

import Foundation

struct Badge: Identifiable, Hashable {
    let id: String
    let emoji: String
    let title: String
    let subtitle: String
}

enum Badges {
    static let catalog: [Badge] = [
        Badge(id: "first_night", emoji: "🌙", title: "First Night", subtitle: "You logged your first session"),
        Badge(id: "first_shot", emoji: "🥃", title: "First Shot", subtitle: "One shot down"),
        Badge(id: "hydration_hero", emoji: "💧", title: "Hydration Hero", subtitle: "3+ waters in one night"),
        Badge(id: "home_before_midnight", emoji: "🕛", title: "Home by Midnight", subtitle: "Wrapped before 12am"),
        Badge(id: "marathoner", emoji: "🏃", title: "Marathoner", subtitle: "10+ drinks in a night"),
        Badge(id: "in_the_zone", emoji: "🎯", title: "In the Zone", subtitle: "Nailed your target with no overshoot"),
        Badge(id: "designated_driver", emoji: "🚗", title: "Designated Driver", subtitle: "Zero drinks, still out with the squad"),
        Badge(id: "streak_5", emoji: "🔥", title: "5-Night Streak", subtitle: "Five nights logged in a row"),
        Badge(id: "sommelier", emoji: "🍷", title: "Sommelier", subtitle: "Wine-only night"),
        Badge(id: "brewmaster", emoji: "🍺", title: "Brewmaster", subtitle: "Beer-only night"),
        Badge(id: "mixologist", emoji: "🍹", title: "Mixologist", subtitle: "Cocktail-only night"),
        Badge(id: "explorer", emoji: "🌍", title: "Explorer", subtitle: "All 4 drink types in one night"),
    ]

    private static let byId: [String: Badge] =
        Dictionary(uniqueKeysWithValues: catalog.map { ($0.id, $0) })

    static func lookup(_ id: String) -> Badge? { byId[id] }

    /// Evaluate all badges against the just-ended session (`session`) and the
    /// full history including it. Returns badges newly earned.
    static func evaluate(session: NightSession, profile: Profile, history: [NightSession]) -> [Badge] {
        var earned: [Badge] = []

        // First night: always awarded when history has at least one entry.
        if history.count >= 1 { earned.append(byId["first_night"]!) }

        // First shot ever.
        let anyShot = history.contains { s in s.drinks.contains { $0.type == .shot } }
        if anyShot { earned.append(byId["first_shot"]!) }

        // Session-specific badges.
        if session.waters.count >= 3 { earned.append(byId["hydration_hero"]!) }

        if let end = session.endedAt, !session.drinks.isEmpty {
            let cal = Calendar.current
            let startDay = cal.startOfDay(for: session.startedAt)
            let midnight = cal.date(byAdding: .day, value: 1, to: startDay) ?? end
            if end < midnight {
                earned.append(byId["home_before_midnight"]!)
            }
        }

        if session.drinks.count >= 10 { earned.append(byId["marathoner"]!) }

        // Target hit cleanly.
        let target = TargetsCatalog.target(session.targetId)
        let end = session.endedAt ?? Date()
        var peak = 0.0
        var everInZone = false
        var everOver = false
        var t = session.startedAt
        while t <= end {
            let b = BAC.estimate(drinks: session.drinks, profile: profile, at: t)
            peak = max(peak, b)
            if b >= target.minBac && b <= target.maxBac { everInZone = true }
            if b > target.maxBac { everOver = true }
            t = t.addingTimeInterval(5 * 60)
        }
        if everInZone && !everOver && !session.drinks.isEmpty {
            earned.append(byId["in_the_zone"]!)
        }

        if session.drinks.isEmpty && (session.endedAt?.timeIntervalSince(session.startedAt) ?? 0) > 45 * 60 {
            earned.append(byId["designated_driver"]!)
        }

        // 5-night streak - trailing 5 sessions each on different days.
        let recentDays = history.suffix(5).map { Calendar.current.startOfDay(for: $0.startedAt) }
        if recentDays.count == 5, Set(recentDays).count == 5 {
            earned.append(byId["streak_5"]!)
        }

        let uniqueTypes = Set(session.drinks.map(\.type))
        if uniqueTypes == [.wine] { earned.append(byId["sommelier"]!) }
        if uniqueTypes == [.beer] { earned.append(byId["brewmaster"]!) }
        if uniqueTypes == [.cocktail] { earned.append(byId["mixologist"]!) }
        if uniqueTypes.count == 4 { earned.append(byId["explorer"]!) }

        return earned
    }
}

File: beerify/src/lib/badges.ts
import type { NightSession, Profile } from '../types'
import { estimateBac } from './bac'
import { TARGETS } from './drinks'

export interface Badge {
  id: string
  emoji: string
  title: string
  subtitle: string
}

export const BADGES: Badge[] = [
  { id: 'first-night', emoji: '🌙', title: 'First Night', subtitle: 'Log your first session' },
  { id: 'first-shot', emoji: '🥃', title: 'First Shot', subtitle: 'Log a shot' },
  { id: 'hydration-hero', emoji: '💧', title: 'Hydration Hero', subtitle: 'Three waters in one night' },
  { id: 'home-before-midnight', emoji: '🕛', title: 'Home by Midnight', subtitle: 'Wrap a drinking night before midnight' },
  { id: 'marathoner', emoji: '🏃', title: 'Marathoner', subtitle: 'Ten drinks in one night' },
  { id: 'in-the-zone', emoji: '🎯', title: 'In the Zone', subtitle: 'Hit the target without overshooting' },
  { id: 'designated-driver', emoji: '🚗', title: 'Designated Driver', subtitle: 'Go 45 minutes with zero alcohol' },
  { id: 'streak-5', emoji: '🔥', title: '5-Night Streak', subtitle: 'Log five consecutive calendar days' },
  { id: 'sommelier', emoji: '🍷', title: 'Sommelier', subtitle: 'Wine only for a night' },
  { id: 'brewmaster', emoji: '🍺', title: 'Brewmaster', subtitle: 'Beer or cider only for a night' },
  { id: 'mixologist', emoji: '🍹', title: 'Mixologist', subtitle: 'Cocktails only for a night' },
  { id: 'explorer', emoji: '🌍', title: 'Explorer', subtitle: 'Four drink families in one night' },
]

function hasFiveDayStreak(history: NightSession[]): boolean {
  const days = [...new Set(history.map((session) => new Date(new Date(session.startedAt).setHours(0, 0, 0, 0)).getTime()))].sort((a, b) => a - b)
  let streak = 1
  for (let index = 1; index < days.length; index++) {
    streak = days[index] - days[index - 1] === 86_400_000 ? streak + 1 : 1
    if (streak >= 5) return true
  }
  return false
}

function hitTargetCleanly(session: NightSession, profile: Profile): boolean {
  if (!session.drinks.length) return false
  const target = TARGETS[session.targetId]
  const end = session.endedAt ?? session.updatedAt
  let hit = false
  for (let at = session.startedAt; at <= end; at += 5 * 60_000) {
    const bac = estimateBac(session.drinks, profile, at, session.mealState)
    if (bac > target.maxBac) return false
    if (bac >= target.minBac) hit = true
  }
  return hit
}

export function earnedBadgeIds(history: NightSession[], profile: Profile): Set<string> {
  const earned = new Set<string>()
  if (history.length) earned.add('first-night')
  if (history.some((session) => session.drinks.some((drink) => drink.category === 'shot'))) earned.add('first-shot')
  if (history.some((session) => session.waters.length >= 3)) earned.add('hydration-hero')
  if (history.some((session) => {
    if (!session.drinks.length || !session.endedAt) return false
    const midnight = new Date(session.startedAt); midnight.setHours(24, 0, 0, 0)
    return session.endedAt < midnight.getTime()
  })) earned.add('home-before-midnight')
  if (history.some((session) => session.drinks.length >= 10)) earned.add('marathoner')
  if (history.some((session) => hitTargetCleanly(session, profile))) earned.add('in-the-zone')
  if (history.some((session) => session.participationMode === 'driver' && !session.drinks.length && (session.endedAt ?? session.updatedAt) - session.startedAt >= 45 * 60_000)) earned.add('designated-driver')
  if (hasFiveDayStreak(history)) earned.add('streak-5')
  for (const session of history) {
    const categories = new Set(session.drinks.map((drink) => drink.category))
    if (categories.size === 1 && categories.has('wine')) earned.add('sommelier')
    if (categories.size > 0 && [...categories].every((category) => category === 'beer' || category === 'cider')) earned.add('brewmaster')
    if (categories.size === 1 && categories.has('cocktail')) earned.add('mixologist')
    const families = new Set(session.drinks.map((drink) => drink.category === 'cider' ? 'beer' : drink.category === 'shot' ? 'spirit' : drink.category))
    if (families.size >= 4) earned.add('explorer')
  }
  return earned
}
